File: learning_tf/src/turtle_tf_broadcaster.cpp
#include <ros/ros.h>
#include <tf/transform_broadcaster.h> // tfのBroadcasterを使うヘッダーファイル読み込み
#include <turtlesim/Pose.h>           // turtlesimで使われるPoseメッセージファイルを読み込み

/**
 * turtleのtfをbroadcastするサンプル(C++)
 */

std::string turtle_name; // 亀の名前を保持する変数

// Subscribeした時のコールバック関数
void poseCallback(const turtlesim::PoseConstPtr& msg){
  // Broadcaster
  static tf::TransformBroadcaster br;

  // tfの3次元座標を作成
  tf::Transform transform;
  transform.setOrigin( tf::Vector3(msg->x, msg->y, 0.0) );

  // tfの姿勢座標を作成
  tf::Quaternion q;
  q.setRPY(0, 0, msg->theta);
  transform.setRotation(q);

  // worldフレームにtfを配信
  // 「ros::Time::now()」でタイムスタンプとして現在のROSの時間を設定
  // 親フレーム：world、子フレーム：変数turtle_nameの値
  br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "world", turtle_name));
}

// メイン関数
int main(int argc, char** argv){
  // ノード名「my_tf_broadcaster」
  ros::init(argc, argv, "my_tf_broadcaster");

  // 引数チェック
  if (argc != 2) { ROS_ERROR("need turtle name as argument"); return -1; };

  // 引数の1つ目を亀の名前にする
  turtle_name = argv[1];

  // 「亀の名前/pose」のトピックをSubscribe
  ros::NodeHandle node;
  ros::Subscriber sub = node.subscribe(turtle_name+"/pose", 10, &poseCallback);

  ros::spin();
  return 0;
};

File: learning_tf/scripts/turtle_tf_broadcaster.py
#!/usr/bin/python3
import roslib
roslib.load_manifest('learning_tf')
import rospy

import tf
import turtlesim.msg

"""
turtleのtfをbroadcastするサンプル(python)
"""

# Subscribeした際のコールバック関数
def handle_turtle_pose(msg, turtlename):
  # Broadcaster
  br = tf.TransformBroadcaster()

  # eulerからquaternionを作成
  q = tf.transformations.quaternion_from_euler(0, 0, msg.theta)

  # worldフレームにtfを配信
  # 「rospy.Time.now()」でタイムスタンプとして現在のROSの時間を設定
  # 親フレーム：world、子フレーム：変数turtlenameの値
  br.sendTransform((msg.x, msg.y, 0), q, rospy.Time.now(), turtlename, "world")


if __name__ == '__main__':
  # ノード名「turtle_tf_broadcaster」
  rospy.init_node('turtle_tf_broadcaster')

  # 引数「turtle」の値を取得
  turtlename = rospy.get_param('~turtle')

  # turtleのPoseを取得してコールバック関数である「handle_turtle_pose」を実行
  rospy.Subscriber('/%s/pose' % turtlename, turtlesim.msg.Pose, handle_turtle_pose, turtlename)
  rospy.spin()
